File: apiClient.java
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

public class ApiClient {
    
    private static final String BASE_URL = "https://api.fa-piao.com";
    private static final String APP_KEY = "YOUR_APP_KEY";
    private static final String APP_SECRET = "YOUR_SECRET_KEY";
    
    public static void main(String[] args) throws Exception {
        String path = "/v5/enterprise/authorization";
        String method = "POST";
        
        // 请求参数
        Map<String, String> formData = new HashMap<>();
        formData.put("nsrsbh", "915101820724315989");
        
        // 生成签名参数
        String randomString = generateRandomString(20);
        String timestamp = String.valueOf(Instant.now().getEpochSecond());
        
        // 计算签名
        String signature = calculateSignature(method, path, randomString, timestamp);
        
        // 构建请求体
        String boundary = "----JavaHttpBoundary";
        String requestBody = buildMultipartBody(formData, boundary);
        
        // 发送请求
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(BASE_URL + path))
                .header("AppKey", APP_KEY)
                .header("Sign", signature)
                .header("TimeStamp", timestamp)
                .header("RandomString", randomString)
                .header("Content-Type", "multipart/form-data; boundary=" + boundary)
                .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                .build();

        // 打印调试信息
        System.out.println(method + ": " + BASE_URL + path);
        System.out.println("headers: " + getHeadersString(signature, timestamp, randomString, boundary));
        System.out.println("form-data: " + formData);

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        
        System.out.println("HTTP状态码: " + response.statusCode());
        System.out.println("响应内容: " + response.body());
    }
    
    private static String generateRandomString(int length) {
        SecureRandom random = new SecureRandom();
        String chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        StringBuilder sb = new StringBuilder();
        
        for(int i=0; i<length; i++) {
            int index = random.nextInt(chars.length());
            sb.append(chars.charAt(index));
        }
        return sb.toString();
    }
    
    private static String calculateSignature(String method, String path, 
                                            String randomString, String timestamp) 
                                            throws Exception {
        String content = String.format(
            "Method=%s&Path=%s&RandomString=%s&TimeStamp=%s&AppKey=%s",
            method, path, randomString, timestamp, APP_KEY
        );
        
        Mac sha256 = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKey = new SecretKeySpec(APP_SECRET.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        sha256.init(secretKey);
        
        byte[] hash = sha256.doFinal(content.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(hash).toUpperCase();
    }
    
    private static String bytesToHex(byte[] hash) {
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }
    
    private static String buildMultipartBody(Map<String, String> data, String boundary) {
        StringBuilder body = new StringBuilder();
        
        for (Map.Entry<String, String> entry : data.entrySet()) {
            body.append("--").append(boundary).append("\r\n")
                .append("Content-Disposition: form-data; name=\"")
                .append(entry.getKey()).append("\"\r\n\r\n")
                .append(entry.getValue()).append("\r\n");
        }
        body.append("--").append(boundary).append("--\r\n");
        
        return body.toString();
    }
    
    private static String getHeadersString(String signature, String timestamp, 
                                         String randomString, String boundary) {
        return String.format(
            "{AppKey=%s, Sign=%s, TimeStamp=%s, RandomString=%s, Content-Type=%s}",
            APP_KEY, signature, timestamp, randomString, 
            "multipart/form-data; boundary=" + boundary
        );
    }
}

File: apiClient.go
package main

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json" // 新增导入
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"strings"
	"time"
)

const (
	BaseURL   = "https://api.fa-piao.com"
	AppKey    = "YOUR_APP_KEY"    // 替换为您的AppKey
	AppSecret = "YOUR_SECRET_KEY" // 替换为您的密钥
)

// 生成随机字符串
func generateRandomString(length int) string {
	const charset = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	b := make([]byte, length)
	rand.Read(b) // 安全随机数生成
	for i := range b {
		b[i] = charset[b[i]%byte(len(charset))]
	}
	return string(b)
}

// 计算HMAC-SHA256签名
func calculateSignature(method, path, randomString, timestamp string) string {
	signContent := fmt.Sprintf(
		"Method=%s&Path=%s&RandomString=%s&TimeStamp=%s&AppKey=%s",
		method, path, randomString, timestamp, AppKey,
	)

	mac := hmac.New(sha256.New, []byte(AppSecret))
	mac.Write([]byte(signContent))
	return strings.ToUpper(hex.EncodeToString(mac.Sum(nil)))
}

// 创建multipart/form-data请求体
func createRequestBody(params map[string]string) (io.Reader, string) {
	body := &strings.Builder{}
	writer := multipart.NewWriter(body)

	for key, value := range params {
		_ = writer.WriteField(key, value)
	}
	writer.Close()

	return strings.NewReader(body.String()), writer.FormDataContentType()
}

func main() {
	// 准备请求参数
	path := "/v5/enterprise/authorization"
	method := "POST"
	params := map[string]string{
		"nsrsbh": "915101820724315989"
	}

	// 生成签名参数
	randomString := generateRandomString(20)
	timestamp := fmt.Sprintf("%d", time.Now().Unix())

	// 计算签名
	signature := calculateSignature(method, path, randomString, timestamp)

	// 创建请求体
	payload, contentType := createRequestBody(params)

	// 创建HTTP请求
	req, _ := http.NewRequest(method, BaseURL+path, payload)

	// 设置请求头
	req.Header.Set("AppKey", AppKey)
	req.Header.Set("Sign", signature)
	req.Header.Set("TimeStamp", timestamp)
	req.Header.Set("RandomString", randomString)
	req.Header.Set("Content-Type", contentType)

	// 打印请求信息
	fmt.Printf("%s: %s%s\n", method, BaseURL, path)

	// 转换headers为可打印格式
	headers := map[string]string{
		"AppKey":       AppKey,
		"Sign":         signature,
		"TimeStamp":    timestamp,
		"RandomString": randomString,
		"Content-Type": contentType,
	}
	headersJson, _ := json.Marshal(headers)
	fmt.Printf("headers: %s\n", headersJson)

	// 打印form-data参数
	paramsJson, _ := json.Marshal(params)
	fmt.Printf("form-data: %s\n", paramsJson)

	// 发送请求
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("请求失败: %v\n", err)
		return
	}
	defer resp.Body.Close()

	// 读取响应
	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("HTTP状态码: %d\n", resp.StatusCode)
	fmt.Printf("响应内容: %s\n", string(body))
}
